src/algebra/category/Module/kernels.lean

```diff
@@ -0,0 +1,64 @@
+/-
+Copyright (c) 2020 Markus Himmel. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Markus Himmel
+-/
+import algebra.category.Module.basic
+
+/-!
+# The concrete (co)kernels in the category of modules are (co)kernels in the categorical sense.
+-/
+
+open category_theory
+open category_theory.limits
+open category_theory.limits.walking_parallel_pair
+
+universe u
+
+namespace Module
+variables {R : Type u} [ring R]
+
+section
+variables {M N : Module R} (f : M ⟶ N)
+
+/-- The kernel cone induced by the concrete kernel. -/
+def kernel_cone : kernel_fork f :=
+kernel_fork.of_ι (as_hom f.ker.subtype) $ by tidy
+
+/-- The kernel of a linear map is a kernel in the categorical sense. -/
+def kernel_is_limit : is_limit (kernel_cone f) :=
+fork.is_limit.mk _
+  (λ s, linear_map.cod_restrict f.ker (fork.ι s) (λ c, linear_map.mem_ker.2 $
+    by { rw [←@function.comp_apply _ _ _ f (fork.ι s) c, ←coe_comp, fork.condition,
+      has_zero_morphisms.comp_zero (fork.ι s) N], refl }))
+  (λ s, linear_map.subtype_comp_cod_restrict _ _ _)
+  (λ s m h, linear_map.ext $ λ x, subtype.ext_iff_val.2 $
+    have h₁ : (m ≫ (kernel_cone f).π.app zero).to_fun = (s.π.app zero).to_fun,
+      by { congr, exact h zero },
+    by convert @congr_fun _ _ _ _ h₁ x )
+
+/-- The cokernel cocone induced by the projection onto the quotient. -/
+def cokernel_cocone : cokernel_cofork f :=
+cokernel_cofork.of_π (as_hom f.range.mkq) $ linear_map.range_mkq_comp _
+
+/-- The projection onto the quotient is a cokernel in the categorical sense. -/
+def cokernel_is_colimit : is_colimit (cokernel_cocone f) :=
+cofork.is_colimit.mk _
+  (λ s, f.range.liftq (cofork.π s) $ linear_map.range_le_ker_iff.2 $ cokernel_cofork.condition s)
+  (λ s, f.range.liftq_mkq (cofork.π s) _)
+  (λ s m h,
+  begin
+    haveI : epi (as_hom f.range.mkq) := epi_of_range_eq_top _ (submodule.range_mkq _),
+    apply (cancel_epi (as_hom f.range.mkq)).1,
+    convert h walking_parallel_pair.one,
+    exact submodule.liftq_mkq _ _ _
+  end)
+end
+
+instance : has_kernels (Module R) :=
+⟨λ X Y f, ⟨_, kernel_is_limit f⟩⟩
```

Comment:
Notice that this now gives us potentially two instances of `has_kernels (Module R)`, once here, and once derived from the general `has_limits (Module R)`. (The same would be true for cokernels, we just haven't constructed general colimits yet.)

We have a few options:
1. as discussed at https://github.com/leanprover-community/mathlib/pull/3562#discussion_r461523487, we can solve this after the fact
2. we could postpone this a little by just making this a `def`, so in `abelian.lean` it picks up the generic definition.

For 2. we'd actually need to add 
```

@[priority 100] -- see Note [lower instance priority]
instance has_equalizers_of_has_limits [has_limits C] : has_equalizers C :=
{ has_limits_of_shape := by apply_instance }
@[priority 100] -- see Note [lower instance priority]
instance has_coequalizers_of_has_colimits [has_colimits C] : has_coequalizers C :=
{ has_colimits_of_shape := by apply_instance }
```
in `equalizers.lean`.

Thoughts?

Reply:
I'm inclined to test option 1. Because this is something that I expect we will run into *a lot*. And I'm not sure that careful juggling of `def`s and `instance`s is going to scale.
On the other hand the trouble with 1 starts when the two `limit` constructions are not near the head of the expression, so the `compare_limits` trick will be a lot harder to play...

Reply:
It think we want the instance in `equalizers.lean` in any case (maybe using `has_finite_limits` rather than `has_limits`?). The fact that it's missing is just an oversight.

What I would really like to see is a way to somehow patch the nice kernels into the generic `has_limits` instance. 

Anyway, I'm fine with both options you presented.

Reply:
Actually, the `has_equalizers_of_has_finite_limits` are already present in `finite_limits.lean`, they'd just need to be changed from a `def` to an `instance`.

Reply:
How about we go with option 2, modulo your fix: in this file you make `has_equalizers_of_has_finite_limits` a local instance. Does that work?

Reply:
@semorrison I've just pushed something. Is this what you had in mind?

Reply:
Thanks!